docs/current/guides/snippets/build-test-publish-php/dagger.php

```diff
@@ -0,0 +1,277 @@
+        withEnvVariable(name: "APP_DEBUG", value: "false") {
+          withEnvVariable(name: "APP_NAME", value: "Laravel with Dagger") {
+            withEntrypoint(args: "/var/www/docker-entrypoint.sh") {
+              id
+            }
+          }
+        }
+      }
+    }
+    QUERY;
+    $appProduction = $this->executeQuery($appProductionQuery);
+    return $appProduction;
+  }
+
+  // run unit tests
+  public function testImage($image) {
+    // create database service container
+    $dbQuery = <<<QUERY
+    query {
+      container {
+        from(address: "mariadb:10.11.2") {
+          withEnvVariable(name: "MARIADB_DATABASE", value: "t_db") {
+            withEnvVariable(name: "MARIADB_USER", value: "t_user") {
+              withEnvVariable(name: "MARIADB_PASSWORD", value: "t_password") {
+                withEnvVariable(name: "MARIADB_ROOT_PASSWORD", value: "root") {
+                  withExposedPort(port: 3306) {
+                    withExec(args: []) {
+                      id
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    QUERY;
+    $db = $this->executeQuery($dbQuery);
+
+    // bind database service to application image
+    // set database credentials for application
+    // run all PHPUnit tests
+    $testQuery = <<<QUERY
+    query {
+      container (id: "$image") {
+        withServiceBinding(alias: "mariadb", service: "$db") {
+          withEnvVariable(name: "DB_HOST", value: "mariadb") {
+            withEnvVariable(name: "DB_USERNAME", value: "t_user") {
+              withEnvVariable(name: "DB_PASSWORD", value: "t_password") {
+                withEnvVariable(name: "DB_DATABASE", value: "t_db") {
+                  withWorkdir(path: "/var/www") {
+                    withExec(args: ["./vendor/bin/phpunit", "-vv"]) {
+                      stdout
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    QUERY;
+    $test = $this->executeQuery($testQuery);
+    return $test;
+  }
+
+  // publish image to registry
+  public function publishImage($image) {
+    // retrieve registry credentials from host environment
+    $registryUsername = getenv("REGISTRY_USERNAME");
+    $registryPassword = getenv("REGISTRY_PASSWORD");
+
+    // set registry password as Dagger secret
+    $registryPasswordSecretQuery = <<<QUERY
+    query {
+      setSecret(name: "password", plaintext: "$registryPassword") {
+        id
+      }
+    }
+    QUERY;
+    $registryPasswordSecret = $this->executeQuery($registryPasswordSecretQuery);
+
+    // authenticate to registry
+    // publish image
+    $publishQuery = <<<QUERY
+    query {
+      container (id: "$image") {
+        withRegistryAuth(address: "docker.io", username: "$registryUsername", secret: "$registryPasswordSecret") {
+          publish(address: "$registryUsername/laravel-dagger")
```

Comment:
I would extract the `address` value into an env var which defaults to `"$registryUsername/laravel-dagger"` if unset.

Reply:
Done